lib/crypt-md5.c

```diff
@@ -83,12 +85,18 @@ crypt_md5crypt_rn (const char *phrase, size_t phr_size,
     /* Skip salt prefix.  */
     salt += sizeof (md5_salt_prefix) - 1;
 
-  salt_size = strspn (salt, b64t);
-  if (salt[salt_size] && salt[salt_size] != '$')
+  /* Find the next '$', or string-terminator if not present. */
+  salt_size = strcspn (salt, "$");
+
+  /* Check if there is no separator for columns ':' in the salt. */
+  badsalt = strpbrk (setting, ":");
+  if (badsalt && badsalt < salt + salt_size)
     {
       errno = EINVAL;
       return;
     }
```

Comment:
I haven't tested this idea, but I believe this sequence could be expressed with just one call to `strcspn`:

```
/* The salt ends at the next '$' or the end of the string.
   Ensure ':' does not appear in the salt (it is used as a separator in /etc/passwd). */
salt_size = strcspn (salt, "$:");
if (salt[salt_size] == ':')
  {
    errno = EINVAL;
    return;
  }
```

The `badsalt` variable then becomes unnecessary.

Reply:
Tested and works.  Will update with next push.